### DISCUSSION TITLE
:class object syntax no longer applies multiple classes under single condition unlike v2

### DISCUSSION THREAD

**Discussion by blackmagic0** [alpinejs/alpine#1525] [7. Bugs (archived)] [answered]
The ':class' attribute no longer applies multiple classes under a single condition when the object syntax is used, only the last class in the string.

This example will result in: ( bg-red-100/bg-blue-100 doesn't get applied )
`<div class="w-12 h-12 border border-red-500'"></div>`
```
<div x-data="example">
    <div class="w-12 h-12 border" :class="{'bg-red-100 border-red-500': isOpen, 'bg-blue-100 border-blue-500':!isOpen}"></div>
    <button @click="toggle">Toggle</button>
</div>
<script>
    document.addEventListener('alpine:initializing', () => {
        Alpine.data('example', () => ({
                isOpen: false,
                toggle: function() {
                    this.isOpen = !this.isOpen;
                }
            })
        );
    })
</script>
```

This is not an issue if you use the non-object syntax:
`<div class="w-12 h-12 border" :class="isOpen ? 'bg-red-100 border-red-500' : 'bg-blue-100 border-blue-500'"></div>`

**Comment by HugoDF** [comment]
This might be an intended breaking change.

You can use the `classnames` package to get the object syntax working again.

**Reply by danddanddand** [reply]
Seems like a very odd choice of this isn't a bug. Even something as simple as toggling a icon requires two classes (font awesome "fal fa-blah" to "far fa-blah2" for example). And with popularity of utility classes in libs like tailwind it would make object syntax almost useless.

**Reply by danddanddand** [reply]
> 
> You can use the `classnames` package to get the object syntax working again.

Can you explain a bit more what you're referring to please?

**Reply by blackmagic0** [reply]
I think that @HugoDF may be referring to https://github.com/JedWatson/classnames

**Reply by HugoDF** [reply]
Yeah that's the one

It's one for @calebporzio it either was an oversight and we can PR a fix or we might want to update the upgrade guide?

**Comment by KevinBatdorf** [comment, marked as answer]
So I actually can't reproduce this. Can you share a demo of it breaking? There may be some context missing.

Here's a demo of it working: https://codepen.io/KevinBatdorf/pen/qBrvOVe?editors=1010

**Reply by blackmagic0** [reply]
Your example worked without issue, I then made my own codepen example that is the same code again ~~and it is broken... don't really know what to think about that.  does my codepen example work correctly for you?~~

it looks like in mine it does work I had a typo where both conditions of :class had the same 'bg-red-100' instead of one having 'bg-blue-100'.. ( typo left in my codepen to demonstrate )

I am unsure why both conditional classes having the same class would cause neither to take affect...

https://codepen.io/jacobpozaic/pen/GRWeZvL

on my codepen:
![165547](https://user-images.githubusercontent.com/25399162/122471163-e9370400-cf8c-11eb-97b4-071db0ea9387.png)

on your codepen:
![165649](https://user-images.githubusercontent.com/25399162/122471289-1388c180-cf8d-11eb-9e9b-96c356cc98a3.png)

**Reply by ryangjchandler** [reply]
@JacobPozaic if both conditions have the same class, Alpine will first say that the class needs to be added. It then checks all of the false conditions to see which classes need to be removed, therefore removing the same class that was already added.

**Reply by blackmagic0** [reply]
Okay, thanks @ryangjchandler!

**Reply by danddanddand** [reply]
> if both conditions have the same class, Alpine will first say that the class needs to be added. It then checks all of the false conditions to see which classes need to be removed, therefore removing the same class that was already added.

@ryangjchandler doesn't this seem very fragile? For example in your code pen if you move border into both conditions, you just get no borders

**Reply by KevinBatdorf** [reply]
> does my codepen example work correctly for you?

Your pen works for me as well. Not sure what you could try.

@ryangjchandler I think you commented on the wrong thread and meant to reply to @danddanddand

**Comment by danddanddand** [comment]
Doesn't it seem wrong that this worked in v2 and not v3? Cases where classes like icons come in from server side rendering and class="{'fal fa-icon':isOpen, 'fas fa-icon2':!isOpen}" works but  class="{'fas fa-icon':isOpen, 'fas fa-icon2':!isOpen}" doesn't work just because one is light family and one is solid family.

```
<div class="p-16">
<div x-data="{
                isOpen: false,
                toggle: function() {
                    this.isOpen = !this.isOpen;
                }
            }">
    <div class="w-12 h-12" :class="{'border bg-red-100 border-red-500': isOpen, 'border bg-blue-100 border-blue-500':!isOpen}"></div>
    <button @click="toggle">Toggle</button>
</div>
</div>
```

**Reply by KevinBatdorf** [reply]
I agree there's room for improvement here. Here's where you would want to PR an optimization. Maybe just by switching line 54 and 53 at the end. You would want to write a test though as well.

https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/utils/classes.js#L29

**Reply by danddanddand** [reply]
Thanks @KevinBatdorf. So the original classList is memoized and it's applying classes to any true condition, then removing any classes to the false conditions in the object keys?

I made a codepen to try to wrap my head around what it's doing (https://codepen.io/danddanddand/pen/ExWOavo) and I get it technically, I just am trying to understand @calebporzio's intent. I think v2 worked by memoizing elements original class and only applying classes to true conditions of the object's keys. You can mess around with the codepen and see it's hard to guess what will be added/removed to classList from glancing at code (at least for me). Maybe we were using v2 wrong. I'm struggling to understand it so will transition our code to ternary expressions for now.

**Reply by KevinBatdorf** [reply]
I don't think it's helpful to tag Caleb in these discussions. Better to prepare a PR with a fix or even a failing test case.

V3 is currently adding the truthy values first, then removing falsy values. I think the logic should be reversed so you end up with something like this (see pen below). Although I'm not entirely sure if it will affect anything else (it's not clear to me what the return function does in the code I linked above)

I modified V3 and added it to your codepen. i'm assuming this is how you expect it to work?
https://codepen.io/KevinBatdorf/pen/VwpRxMb

**Reply by danddanddand** [reply]
Thanks @KevinBatdorf . Sorry about tagging Caleb. 

Yes, this example works how V2 works. I had to fix some code since I did it on my phone last night but this is your codepen forked with v2 instead of v3 (https://codepen.io/danddanddand/pen/yLMwEBN). You can see border is actually applied in case 1/5 so I'll look at making PR.

**Reply by KevinBatdorf** [reply]
If you can't work out the PR let me know and I can submit one maybe later or tomorrow

**Reply by danddanddand** [reply]
I have co-worker doing it, I suspect he should have something later today for review.

**Reply by blackmagic0** [reply]
https://github.com/alpinejs/alpine/pull/1556
